### PR TITLE
feat(scraper-proxy): support query cursors

### PR DESCRIPTION
**Pull request by troykessler**
### Description

Adds cursor pagination to `message_view` GraphQL queries. A message cursor must contain exactly one non-null `id`. `order_by` may be omitted; when omitted, the cursor direction determines the ordering. When provided, its single effective column and direction must match the cursor.

### Drive-by changes

- Preserves special GraphQL variable keys during cache-key canonicalization.
- Keeps nested catch-up failure details in server logs and returns a stable client-safe WebSocket error.
- Adds lightweight utils subpath exports to avoid loading the root barrel in scraper-proxy.

### Related issues

None.

### Backward compatibility

Yes. The cursor argument is optional; existing queries are unchanged.

### Testing

- `pnpm -C typescript/utils test` (353 tests)
- `pnpm -C typescript/utils lint`
- `pnpm -C typescript/utils build`
- `pnpm -C typescript/scraper-proxy test` (54 tests)
- `pnpm -C typescript/scraper-proxy lint`
- `pnpm -C typescript/scraper-proxy build`
- `pnpm exec changeset status`

### CODE REVIEW

File: typescript/scraper-proxy/src/scraperdb/sql.ts [hunk 1]
```diff
@@ -309,6 +310,26 @@ function validate(args: SelectArgs): void {
   validateWhere(args.where, 0, state);
 }
 
+function validateCursorOrder(
+  cursors: Cursor[] | undefined,
+  orders: Order[],
+): void {
+  if (!cursors?.length || !orders.length) return;
```
Comment (paulbalaji) [inline typescript/scraper-proxy/src/scraperdb/sql.ts]: Low: require exactly one non-null cursor value. With no `order_by`, `cursor: [{ initial_value: {} }]` returns here; `cursorWhere()` and the cursor-derived `orderBy()` then both emit nothing, so the request silently becomes page 1 again instead of failing. This can loop or duplicate a consumer that treats the accepted cursor as a continuation token. Reject empty cursor objects (alongside the existing maximum-one check) and add an executed GraphQL/resolver regression.

**Review by paulbalaji — COMMENTED** [review]
Reviewed at `b6832bbd893b74cc73877f3bc4319e787daa199e`.

I independently reproduced the two unresolved data-integrity findings already raised at `sql.ts:327`: a single-column strict cursor skips tied rows on non-unique fields, and nullable/null cursor values make parts of the ordered result unreachable. Those are core pagination-correctness issues, so I cannot approve this head. Restrict query cursors to a proven unique, non-null key, or implement a NULL-aware lexicographic `(sort_key, unique_key)` cursor.

I did not find SQL injection: cursor identifiers are allowlisted/quoted and values are parameterized. The global-limit compatibility concern at `sql.ts:42` does not appear externally breaking because the served schema strips `subscription_root`, only `message_view` gains a query cursor here, and there are no in-repo cursor callers.

Exact-head verification: 52/52 scraper-proxy tests, build, lint, changeset status, `git diff --check`, and the mandatory cast/`any` audit passed. The new schema test only validates SDL acceptance; it does not execute GraphQL through the resolver or prove multi-page completeness. CI was still running at submission.

File: typescript/scraper-proxy/src/scraperdb/sql.ts [hunk 2]
```diff
@@ -39,7 +39,7 @@ export interface SelectArgs {
 const MAX = {
   boolDepth: 8,
   boolPredicates: 100,
-  cursorColumns: 3,
+  cursorColumns: 1,
```
[the remark this answers is not in the document]
Comment (troykessler) [reply typescript/scraper-proxy/src/scraperdb/sql.ts]: Fixed in e28ebb86ee. Validation is table-aware: message_view requires one id cursor, while domain and raw_message_dispatch retain the existing three-column limit. Added a compatibility test for a two-column domain cursor.

File: typescript/scraper-proxy/src/scraperdb/sql.ts [hunk 3]
```diff
@@ -309,6 +310,26 @@ function validate(args: SelectArgs): void {
...
+  const cursor = cursors[0];
+  const cursorColumn = Object.keys(cursor.initial_value)[0];
+  const orderColumns = orders
+    .flatMap(Object.entries)
+    .filter((entry) => entry[1] != null);
+  const [orderColumn, orderDirection] = orderColumns[0] ?? [];
+  if (
+    orderColumns.length !== 1 ||
+    cursorColumn !== orderColumn ||
+    (cursor.ordering === 'DESC') !== orderDirection?.startsWith('desc')
```
[the remark this answers is not in the document]
Comment (troykessler) [reply typescript/scraper-proxy/src/scraperdb/sql.ts]: Fixed in e28ebb86ee. message_view cursors now accept only id, which is sourced from the underlying message primary key, so pagination order is unique and deterministic. Matching ASC and DESC cases are covered.

File: typescript/scraper-proxy/src/scraperdb/sql.ts (hunk 1 above)
Comment (troykessler) [reply typescript/scraper-proxy/src/scraperdb/sql.ts]: Fixed in e28ebb86ee. message_view now requires exactly one non-null id cursor value. Added an Apollo executeOperation regression proving an empty initial_value fails through the resolver.

File: typescript/scraper-proxy/src/scraperdb/sql.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (troykessler) [reply typescript/scraper-proxy/src/scraperdb/sql.ts]: Fixed in e28ebb86ee. Nullable and non-unique message_view cursor columns are rejected by restricting cursors to id, and null cursor values are rejected before SQL rendering.

File: typescript/scraper-proxy/src/scraperdb/sql.ts [hunk 4]
```diff
@@ -309,6 +310,26 @@ function validate(args: SelectArgs): void {
...
+  ) {
+    throw new Error('cursor columns and directions must match order_by');
```
[the remark this answers is not in the document]
Comment (troykessler) [reply typescript/scraper-proxy/src/scraperdb/sql.ts]: Fixed in e28ebb86ee. Cursor invariants now use assert from @hyperlane-xyz/utils; the package dependency and lockfile were updated.

File: typescript/scraper-proxy/src/scraperdb/cache-plugin.ts [hunk 5]
```diff
@@ -118,24 +119,17 @@ function cacheKey(
   const dataVariables = Object.fromEntries(
     Object.entries(variables).filter(([name]) => usedVariables.has(name)),
   );
+  const serializedVariables = JSON.stringify(sortObjectKeys(dataVariables));
```
Comment (paulbalaji) [inline typescript/scraper-proxy/src/scraperdb/cache-plugin.ts:122]: **Low — preserve special keys in the cache identity.** `sortObjectKeys` assigns keys onto a normal `{}`, so an own `__proto__` GraphQL variable invokes the legacy prototype setter and is omitted from `serializedVariables`. I reproduced this through Apollo: the same cached query with two different `$__proto__` values returned the first result twice, with one resolver call. The removed string serializer preserved the key. Please use prototype-safe canonicalization (for example, null-prototype/own-property construction) and add this regression case.

File: typescript/scraper-proxy/src/scraperdb/sql.ts [hunk 6]
```diff
@@ -297,16 +297,44 @@ function validate(args: SelectArgs): void {
...
+
+  assert(
+    cursors.length === 1 && cursorColumns === 1,
+    'message_view cursor must contain one column',
+  );
+  const cursor = cursors[0];
+  const [cursorColumn, cursorValue] = Object.entries(cursor.initial_value)[0];
+  assert(cursorColumn === 'id', 'message_view cursor column must be id');
+  assert(cursorValue != null, 'message_view cursor value must be non-null');
+  if (!orders.length) return;
```
Comment (paulbalaji) [inline typescript/scraper-proxy/src/scraperdb/sql.ts]: **Low — validate the effective cursor ordering.** With an `id` cursor, `distinct_on: [origin_domain]`, and no explicit `order_by`, this returns early and later emits `DISTINCT ON ("origin_domain") ... ORDER BY "id" ASC`; PostgreSQL rejects it because the distinct expression is not the leftmost ordering. Supplying `order_by: {id: null}` exposes the inverse mismatch: GraphQL accepts it and the renderer would fall back to the cursor order, but validation rejects its zero effective order columns. Please normalize non-null order columns, validate `distinct_on` against the cursor-derived fallback, and cover both compositions.

File: typescript/scraper-proxy/src/live/event-websocket.ts [hunk 7]
```diff
@@ -531,7 +532,7 @@ export class EventWebSocketServer {
       }
       websocketCatchUps.inc({ outcome: 'failure' });
       client.subscriptions.delete(request.eventType);
-      const reason = errorMessage(error);
+      const reason = formatError(error);
```
Comment (paulbalaji) [inline typescript/scraper-proxy/src/live/event-websocket.ts:535]: **Low — keep nested causes server-side.** `reason` is logged but also returned to the unauthenticated WebSocket client below. `formatError` recursively appends `Error.cause`; `pg-pool` wraps connection timeouts with the underlying network error, and a catch-up failure can therefore return internal details such as `connect ECONNREFUSED 10.23.45.67:5432`. Please retain `formatError` for the log but send a stable client-safe error (or only the outer message).

**Review by paulbalaji — COMMENTED** [review]
Full rereview at `d0a45b111b9615fcf1018ef78ddb7fecce000105`. The five earlier threads are resolved/outdated, and I verified their unique/non-null/empty-cursor, table-scoping, assert-style, and positive-test fixes at this head.

I found three low-severity issues in the current diff; details are inline. This is a non-blocking comment review.

Validation: scraper-proxy tests 53/53, build, lint, utils build, `git diff --check`, cast/`any` audit, and current CI passed. Three independent Codex passes and three Kimi passes covered SQL/pagination, server/cache behavior, security, callers, history, and prior discussions.

File: typescript/scraper-proxy/src/scraperdb/sql.ts (hunk 6 above)
Comment (troykessler) [reply typescript/scraper-proxy/src/scraperdb/sql.ts]: Fixed in 1be95e4eca. Validation now removes null order entries first and checks distinct_on against the same cursor-derived fallback used by SQL rendering. Tests cover both omitted order_by and order_by: {id: null}.

File: typescript/scraper-proxy/src/scraperdb/cache-plugin.ts (hunk 5 above)
Comment (troykessler) [reply typescript/scraper-proxy/src/scraperdb/cache-plugin.ts]: Fixed in 1be95e4eca. Cache variable canonicalization now uses Object.fromEntries, which preserves own special keys such as __proto__, and an Apollo regression verifies distinct values produce distinct cache entries.

File: typescript/scraper-proxy/src/live/event-websocket.ts (hunk 7 above)
Comment (troykessler) [reply typescript/scraper-proxy/src/live/event-websocket.ts]: Fixed in 1be95e4eca. Detailed nested errors remain in server logs, while catch-up failures now return only the stable client-safe message. Existing integration cases now assert the sanitized response.

File: typescript/scraper-proxy/package.json [hunk 8]
```diff
@@ -32,6 +32,7 @@
   "dependencies": {
     "@apollo/server": "catalog:",
     "@as-integrations/express5": "catalog:",
+    "@hyperlane-xyz/utils": "workspace:^",
```
Comment (paulbalaji) [inline typescript/scraper-proxy/package.json:35]: **Low — avoid loading the full utils barrel on unbundled service paths.** I measured three clean Node 26 imports of built `scraperdb/tables.js`: base was consistently `0.03s` / ~`50 MB` peak RSS; this head was `0.25–0.35s` / ~`142 MB` because the root ESM barrel evaluates its broad dependency graph. NCC can tree-shake this (the production bundle delta is small), but local `start`, tests, and CI pay the cost. Prefer narrow `@hyperlane-xyz/utils` subpath exports for `assert`, `formatError`, and `setEquality`, or otherwise keep these small helpers off the root barrel. Non-blocking.

File: .changeset/scraper-proxy-query-cursors.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+---
+'@hyperlane-xyz/scraper-proxy': minor
+---
+
+Added cursor pagination to scraper message queries, with validation that cursors use a single column matching the requested sort order.
```
Comment (paulbalaji) [inline .changeset/scraper-proxy-query-cursors.md]: **Low — document the exact public contract.** Current validation accepts only one **non-null `id`** cursor for `message_view`; `order_by` may be omitted, in which case cursor direction supplies the ordering. Please state that here and in the PR description. The PR body also says there are no drive-bys and 52 tests, while this head includes utility/cache/error-response refactors and passes 54 tests.

**Review by paulbalaji — COMMENTED** [review]
Full rereview at exact head `1be95e4ecaed617517a6f26d5d52e06932f5b852` against PR base `309318974f85a5762839fc301a15ab476e017cca`. I reconciled all existing reviews/comments and independently verified all eight resolved threads: the `message_view.id` cursor is unique/non-null and complete in both directions; empty, non-id, mismatched, and incompatible `distinct_on` compositions are rejected; special cache-variable keys remain in the cache identity; and nested catch-up causes stay server-side. I found no remaining cursor correctness, SQL-injection, cache-integrity, or information-disclosure issue.

Two new low-severity, non-blocking observations are inline: the broad utils ESM barrel has a measurable unbundled startup/memory cost, and the public documentation is stale/incomplete. The title remains accurate.

Validation at this head under Node 26.6.0: scraper-proxy tests 54/54, TypeScript build, Oxlint, changeset status, `git diff --check`, mandatory added `as`/`any`/non-null audit, exact SQL edge-case execution, and CI (43 passed / 28 skipped / 0 pending / 0 failed). Current `main` has three disjoint intervening commits and merges cleanly. Independent cursor/security/full-diff agents and three substantive Kimi adversarial passes converged. No live PostgreSQL or production-load validation.

**Review by paulbalaji — APPROVED** [review]
Approved exact head `1be95e4ecaed617517a6f26d5d52e06932f5b852`. All prior cursor-correctness, cache-integrity, and error-disclosure findings are fixed; focused validation and CI are green. The existing utils-barrel performance observation and documentation correction remain non-blocking follow-ups.

File: .changeset/scraper-proxy-query-cursors.md (hunk 9 above)
Comment (troykessler) [reply .changeset/scraper-proxy-query-cursors.md]: Fixed in 220f01c7c6. The changeset now documents exactly one non-null id cursor and the cursor-derived ordering fallback. I also updated the PR description with that contract, the cache/error/utils review fixes, and the current 353 utils and 54 scraper-proxy test counts.

File: typescript/scraper-proxy/package.json (hunk 8 above)
Comment (troykessler) [reply typescript/scraper-proxy/package.json]: Fixed in 220f01c7c6. Added lightweight errors, sets, and validation subpath exports and switched every scraper-proxy utils import to those paths. A clean built tables.js import now completes in 0.02s locally without evaluating the root barrel.

File: typescript/scraper-proxy/src/scraperdb/sql.ts [hunk 10]
```diff
@@ -287,43 +287,66 @@ function validate(args: SelectArgs): void {
...
+  if (!cursors?.length || table !== 'message_view') return;
+
+  assert(
+    cursors.length === 1 && cursorColumns === 1,
+    'message_view cursor must contain one column',
+  );
+  const cursor = cursors[0];
+  const [cursorColumn, cursorValue] = Object.entries(cursor.initial_value)[0];
+  assert(cursorColumn === 'id', 'message_view cursor column must be id');
+  assert(cursorValue != null, 'message_view cursor value must be non-null');
```
Comment (Xaroz) [inline typescript/scraper-proxy/src/scraperdb/sql.ts:325]: **Low — reject malformed PostgreSQL `bigint` cursors at the public boundary.** This only checks non-nullness, while the custom `bigint` scalar passes variable values through unchanged. I reproduced this against the newly pinned `220f01c` production image: `true` reaches PostgreSQL and returns `invalid input syntax for type bigint`, while `"9223372036854775808"` returns `out of range`; both surface as `INTERNAL_SERVER_ERROR`. Because this cursor is the new public continuation boundary used by Explorer, please require a non-negative signed-64-bit integer before rendering SQL and add an executed resolver regression. `parseInteger` in `src/live/protocol.ts` is the closest existing code, but it also needs PostgreSQL's `2^63 - 1` upper bound; I found no suitable validator in utils, SDK, or widgets.

**Review by Xaroz — COMMENTED** [review]
Full rereview at exact head `b61186c7b6e94a351f8e445a243047228437534f` against base `309318974f85a5762839fc301a15ab476e017cca`.

### Result

One new low-severity public-input issue is inline. All 10 prior review threads remain resolved: deterministic unique/non-null `id` pagination, ASC/DESC matching, empty-cursor handling, legacy cursor limits, `distinct_on` composition, special cache-variable keys, client-safe catch-up errors, and lightweight utility imports all hold at this head. I found no SQL-injection, pagination-gap, cache-integrity, WebSocket-disclosure, or utility-compatibility regression.

The new `b61186c` commit only pins mainnet `scraperProxy` to the reviewed `220f01c` image; the tag matches the code under review.

### Verification

- scraper-proxy: 54/54 tests, build, lint, bundle
- utils: 353/353 tests, build, lint
- changeset status and `git diff --check` pass
- direct lightweight import: ~0.03s / 45 MB peak RSS
- production probe against the pinned image reproduced the inline malformed-cursor behavior
- exact-head CI at posting: 23 passed, 59 pending, 0 failed
